internal/moremath/moremath.go

```diff
@@ -186,90 +141,131 @@ func WasmCompatNearestF32(f float32) float32 {
 // the returned NaN value follows the Wasm specification on NaN
 // propagation.
 // https://www.w3.org/TR/2022/WD-wasm-core-2-20220419/exec/numerics.html#nan-propagation
 func WasmCompatCeilF32(f float32) float32 {
-	if math.IsNaN(float64(f)) {
-		return f32ReturnNaNUniOp(math.Float32bits(f))
-	}
-	return float32(math.Ceil(float64(f)))
+	return returnF32UniOp(f, float32(math.Ceil(float64(f))))
 }
 
 // WasmCompatCeilF64 is the same as math.Ceil on 64-bit except that
 // the returned NaN value follows the Wasm specification on NaN
 // propagation.
 // https://www.w3.org/TR/2022/WD-wasm-core-2-20220419/exec/numerics.html#nan-propagation
 func WasmCompatCeilF64(f float64) float64 {
-	if math.IsNaN(f) {
-		return f64ReturnNaNUniOp(math.Float64bits(f))
-	}
-	return math.Ceil(f)
+	return returnF64UniOp(f, math.Ceil(f))
 }
 
 // WasmCompatFloorF32 is the same as math.Floor on 32-bit except that
 // the returned NaN value follows the Wasm specification on NaN
 // propagation.
 // https://www.w3.org/TR/2022/WD-wasm-core-2-20220419/exec/numerics.html#nan-propagation
 func WasmCompatFloorF32(f float32) float32 {
-	if math.IsNaN(float64(f)) {
-		return f32ReturnNaNUniOp(math.Float32bits(f))
-	}
-	return float32(math.Floor(float64(f)))
+	return returnF32UniOp(f, float32(math.Floor(float64(f))))
 }
 
 // WasmCompatFloorF64 is the same as math.Floor on 64-bit except that
 // the returned NaN value follows the Wasm specification on NaN
 // propagation.
 // https://www.w3.org/TR/2022/WD-wasm-core-2-20220419/exec/numerics.html#nan-propagation
 func WasmCompatFloorF64(f float64) float64 {
-	if math.IsNaN(f) {
-		return f64ReturnNaNUniOp(math.Float64bits(f))
-	}
-	return math.Floor(f)
+	return returnF64UniOp(f, math.Floor(f))
 }
 
 // WasmCompatTruncF32 is the same as math.Trunc on 32-bit except that
 // the returned NaN value follows the Wasm specification on NaN
 // propagation.
 // https://www.w3.org/TR/2022/WD-wasm-core-2-20220419/exec/numerics.html#nan-propagation
 func WasmCompatTruncF32(f float32) float32 {
-	if math.IsNaN(float64(f)) {
-		return f32ReturnNaNUniOp(math.Float32bits(f))
-	}
-	return float32(math.Trunc(float64(f)))
+	return returnF32UniOp(f, float32(math.Trunc(float64(f))))
 }
 
 // WasmCompatTruncF64 is the same as math.Trunc on 64-bit except that
 // the returned NaN value follows the Wasm specification on NaN
 // propagation.
 // https://www.w3.org/TR/2022/WD-wasm-core-2-20220419/exec/numerics.html#nan-propagation
 func WasmCompatTruncF64(f float64) float64 {
-	if math.IsNaN(f) {
-		return f64ReturnNaNUniOp(math.Float64bits(f))
+	return returnF64UniOp(f, math.Trunc(f))
+}
+
+func f32IsNaN(v float32) bool {
+	return v != v // this is how NaN is defined.
+}
+
+func f64IsNaN(v float64) bool {
+	return v != v // this is how NaN is defined.
+}
+
+// returnF32UniOp returns the result of 32-bit unary operation. This accepts `original` which is the operand and `result` which is its result,
```

Comment:
nit I've been recently changing docs to 80 column as it is easier to read them without scrolling if you have big fonts.